MyThreadPool/MyThreadPool.Src/Utils.cs

```diff
@@ -0,0 +1,101 @@
+namespace MyThreadPool;
+
+internal class MyThread
+{
+    private readonly MyThreadPool _threadPool;
+    private readonly Thread _thread;
+    public bool Waiting = true;
+    public void Join() => _thread.Join();
+
+    public MyThread(MyThreadPool threadPool)
+    {
+        _threadPool = threadPool;
+        _thread = new(EventLoop);
+        _thread.Start();
+    }
+
+    public void EventLoop()
+    {
+        _threadPool.NewTask.WaitOne();
+        while (true)
+        {
+            if (_threadPool.Queue.TryDequeue(out var task))
+            {
+                Waiting = false;
+                task();
+                continue;
+            }
+            if (_threadPool.TokenSource.IsCancellationRequested)
+            {
+                break;
+            }
+            Waiting = true;
+            _threadPool.NewTask.WaitOne();
+        }
+        Interlocked.Increment(ref _threadPool.ThreadsFinished);
+        _threadPool.NewThreadFinished.Set();
+        _thread.Join();
+    }
+}
+
+internal class MyTask<TResult>(MyThreadPool threadPool, Func<TResult> func) : IMyTask<TResult>
+{
+    private readonly ManualResetEvent _completed = new(false);
+    public bool IsCompleted { get; private set; } = false;
+    private TResult? _result;
+    private readonly Func<TResult> _func = func ?? throw new ArgumentNullException("Function cannot be null");

```

Comment:
```suggestion
    private readonly Func<TResult> _func = func ?? throw new ArgumentNullException(nameof(func));
```
По типу исключения понятно, что пошло не так